Stepic/Modules/CourseInfo/DownloadControlView.swift

```diff
@@ -0,0 +1,222 @@
+        layer.fillColor = nil
+        layer.strokeColor = self.appearance.downloadingCircleColor.cgColor
+        layer.lineWidth = self.appearance.circleWidth
+        layer.strokeEnd = 0
+        return layer
+    }()
+
+    private lazy var downloadingBackgroundCircleLayer: CAShapeLayer = {
+        let layer = CAShapeLayer()
+        layer.fillColor = nil
+        layer.strokeColor = self.appearance.downloadingBackgroundColor.cgColor
+        layer.lineWidth = self.appearance.circleWidth
+        return layer
+    }()
+
+    init(
+        frame: CGRect = .zero,
+        initialState: ActionState = .readyToDownloading,
+        appearance: Appearance = Appearance()
+    ) {
+        self.appearance = appearance
+        self.actionState = initialState
+        super.init(frame: frame)
+
+        self.setupView()
+        self.addSubviews()
+        self.makeConstraints()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        self.updatePendingCirclePath()
+        self.updateDownloadingCirclePath()
+    }
+
+    private func updatePendingCirclePath() {
+        let pendingCirclePath = UIBezierPath()
+        pendingCirclePath.addArc(
+            withCenter: CGPoint(x: self.bounds.midX, y: self.bounds.midY),
+            radius: self.bounds.width / 2 - self.appearance.circleWidth,
+            startAngle: 0,
+            endAngle: 7 * .pi / 4,
+            clockwise: true
+        )
+        self.pendingCircleLayer.path = pendingCirclePath.cgPath
+        self.pendingCircleLayer.frame = self.bounds
+    }
+
+    private func addPendingAnimation() {
+        let rotation = CABasicAnimation(keyPath: "transform.rotation")
+        rotation.byValue = 2 * Float.pi
+        rotation.duration = Animation.pendingDuration
+        rotation.repeatCount = .infinity
+
+        self.pendingCircleLayer.add(rotation, forKey: "circleRotation")
+    }
+
+    private func updateDownloadingCirclePath() {
+        let downloadingCirclePath = UIBezierPath()
+        downloadingCirclePath.addArc(
+            withCenter: CGPoint(x: self.bounds.midX, y: self.bounds.midY),
+            radius: self.bounds.width / 2 - self.appearance.circleWidth,
+            startAngle: -.pi / 2,
+            endAngle: 3 * .pi / 2,
+            clockwise: true
+        )
+
+        self.downloadingBackgroundCircleLayer.path = downloadingCirclePath.cgPath
+        self.downloadingBackgroundCircleLayer.frame = self.bounds
+
+        self.downloadingCircleLayer.path = downloadingCirclePath.cgPath
+        self.downloadingBackgroundCircleLayer.frame = self.bounds
+    }
+
+    private func updateDownloadingProgress(_ progress: Float) {
+        self.downloadingCircleLayer.strokeEnd = CGFloat(progress)
+    }
+
+    private func updateVisibility() {
+        switch self.actionState {
+        case .downloading(_):
```

Comment:
Можно и без параметра, но мне норм

```suggestion
        case .downloading:
```